File: FEM-MATLAB/colsol/colsol.m
function [IERR, K, R] = colsol(n, m, K, R)
% Solve finite element static equilibrium equations in core using
% the active column solver
%
% Input parameters
%   n      - Number of equations
%   m[n]   - Define the skyline of the stiffness matrix K
%            The row number of the first nonzero element in each column
%   K[n,n] - The stiffness matrix
%   R[n]   - Right-hand-side load vector
%
% Output parameters
%   IERR   - Error indicator. If IERR > 0, K is not positive definite.
%   K      - D and L (Factors of stiffness matrix)
%            The elements of D are stored on its diagonal, 
%            and L replaces its upper triangular part
%   R      - Displacement vector
%
%
% Perform L*D*L(T) factorization of stiffness matrix
%   LDLT is an active column solver to obtain the LDLT factorization
%   of a stiffness matrix K

IERR = 0;
for j = 2:n
    
    for i = m(j)+1:j-1
        c = 0.0;
        for r = max(m(i),m(j)):i-1
            c = c + K(r,i)*K(r,j);
        end
        K(i,j) = K(i,j) - c;
    end
    
    for i = m(j):j-1
        Lij = K(i,j)/K(i,i);
        K(j,j) = K(j,j) - Lij*K(i,j);
        K(i,j) = Lij;
    end

    if K(j,j) <= 0
        fprintf('Error - stiffness matrix is not positive definite !\n')
        fprintf('        Nonpositive pivot for equation %d\n', n)
        fprintf('        Pivot = %f\n', K(j,j))
            
        IERR = j;
        return
    end
end

% Reduce right-hand-side load vector
for i = 2:n
    for j = m(i):i-1
        R(i) = R(i) - K(j,i) * R(j);
    end
end

%back-substitute
for i = 1:n
    R(i) = R(i)/K(i,i);
end

for j = n:-1:2
    for i = m(j):j-1
        R(i) = R(i) - K(i,j)*R(j);
    end
end

File: FEM-python/colsol-python/colsol.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
provides colsol function to solve finite element static equilibrium equations 
  in core using the active column solver

Created on Fri Jun 19 13:12:36 2020

@author: xzhang@tsinghua.edu.cn
"""

def colsol(n,m,K,R):
    """
    colsol - Solve finite element static equilibrium equations in core using
      the active column solver
    
    Usage:
        IERR, K, R = colsol(n, m, K, R)
    Input parameters
       n      - Number of equations
       m[n]   - Define the skyline of the stiffness matrix K
                The row number of the first nonzero element in each column
       K[n,n] - The stiffness matrix
       R[n]   - Right-hand-side load vector
    
    Output parameters
       IERR   - Error indicator. If IERR > 0, K is not positive definite.
       K      - D and L (Factors of stiffness matrix)
                The elements of D are stored on its diagonal, 
                and L replaces its upper triangular part
       R      - Displacement vector
    """
    
    IERR = 0

    # Perform L*D*L(T) factorization of stiffness matrix
    # LDLT is an active column solver to obtain the LDLT factorization
    # of a stiffness matrix K
    # Note that all indices are zero-based in Python
    for j in range(1, n):  
        for i in range(m[j]+1, j):
            c = 0.0
            for r in range(max(m[i],m[j]), i):
                c += K[r,i]*K[r,j]

            K[i,j] -= c
        
        for i in range(m[j], j):
            Lij = K[i,j]/K[i,i]
            K[j,j] = K[j,j] - Lij*K[i,j]
            K[i,j] = Lij

        if K[j,j] <= 0:
            print('Error - stiffness matrix is not positive definite !')
            print('        Nonpositive pivot for equation ', j)
            print('        Pivot = ', K[j,j])

            IERR = j
            return IERR, K, R

    # Reduce right-hand-side load vector
    for i in range(1, n):
        for r in range(m[i], i):
            R[i] -= K[r,i] * R[r]
    
    for i in range(0, n):
        R[i] /= K[i,i]

    # Back-substitute
    for i in range(n-1, 0, -1):
        for r in range(m[i], i):
            R[r] -= K[r,i]*R[i]

    return IERR, K, R
